alchemist/alchemist-interfaces/src/main/kotlin/it/unibo/alchemist/model/interfaces/Obstacle.kt

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 2010-2020, Danilo Pianini and contributors
+ * listed in the main project's alchemist/build.gradle.kts file.
+ *
+ * This file is part of Alchemist, and is distributed under the terms of the
+ * GNU General Public License, with a linking exception,
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.model.interfaces
+
+import it.unibo.alchemist.model.interfaces.geometry.Vector
+import java.io.Serializable
+
+/**
+ * A generic obstacle in a vector space.
+ *
+ * @param V the vector type for the space in which this obstacle is placed.
+ */
+interface Obstacle<V : Vector<V>> : Serializable {
```

Comment:
In the first place, I moved all the methods of `Obstacle2D` to `Obstacle` (and of course I used the type parameter). However, I later changed my mind and moved all those methods back to `Obstacle2D`, but kept the type parameter on `Obstacle`. At present, there is no particular reason to maintain it, even though removing it means `EnvironmentWithObstacle` will become an environment with generic obstacles (instead of a vector space with coherent obstacles).

Reply:
Let's reason for a moment on what an `Obstacle` is then. As it stands, an obstacle is "something with an identifier". What do we need out of an Obstacle?

Reply:
An obstacle is an n-dimensional shape providing some specific methods related to the intersection with vectors, which are those used by environments with obstacles. These method are now defined in `Obstacle2D`, I'll move them back to `Obstacle`.

Reply:
Makes sense